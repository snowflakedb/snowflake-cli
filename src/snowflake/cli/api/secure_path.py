# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import errno
import logging
import os
import shutil
import tempfile
from contextlib import contextmanager
from pathlib import Path
from typing import Optional, Union

from snowflake.cli.api.exceptions import DirectoryIsNotEmptyError, FileTooLargeError
from snowflake.cli.api.secure_utils import (
    chmod as secure_chmod,
)
from snowflake.cli.api.secure_utils import (
    restrict_file_permissions,
)

log = logging.getLogger(__name__)

UNLIMITED = -1


class SecurePath:
    def __init__(self, path: Union[Path, str]):
        self._path = Path(path)

    def __repr__(self):
        return f'SecurePath("{self._path}")'

    def __truediv__(self, key):
        return SecurePath(self._path / key)

    @property
    def path(self) -> Path:
        """
        Returns itself in pathlib.Path format
        """
        return self._path

    def chmod(self, permissions_mask: int) -> None:
        """
        Change the file mode and permissions, like os.chmod().
        """
        secure_chmod(self._path, permissions_mask)

    @property
    def parent(self):
        """
        The logical parent of the path. For details, check pathlib.Path.parent
        """
        return SecurePath(self._path.parent)

    def absolute(self):
        """
        Make the path absolute, without normalization or resolving symlinks.
        """
        return SecurePath(self._path.absolute())

    def iterdir(self):
        """
        When the path points to a directory, yield path objects of the directory contents.
        Otherwise, NotADirectoryError is raised.
        If the location does not exist, FileNotFoundError is raised.

        For details, check pathlib.Path.iterdir()
        """
        self.assert_exists()
        self.assert_is_directory()
        return (SecurePath(p) for p in self._path.iterdir())

    def exists(self) -> bool:
        """
        Return True if the path points to an existing file or directory.
        """
        return self._path.exists()

    def is_dir(self) -> bool:
        """
        Return True if the path points to a directory (or a symbolic link pointing to a directory),
        False if it points to another kind of file.
        """
        return self._path.is_dir()

    def is_file(self) -> bool:
        """
        Return True if the path points to a regular file (or a symbolic link pointing to a regular file),
        False if it points to another kind of file.
        """
        return self._path.is_file()

    def glob(self, pattern: str):
        """
        Return a generator yielding Path objects that match the given pattern.
        """
        return self._path.glob(pattern)

    @property
    def name(self) -> str:
        """A string representing the final path component."""
        return self._path.name

    def restrict_permissions(self) -> None:
        """
        Restrict file/directory permissions to owner-only.
        """
        restrict_file_permissions(self._path)

    def touch(self, permissions_mask: int = 0o600, exist_ok: bool = True) -> None:
        """
        Create a file at this given path. For details, check pathlib.Path.touch()
        """
        if not self.exists():
            log.info("Creating file %s", str(self._path))
        self._path.touch(mode=permissions_mask, exist_ok=exist_ok)

    def mkdir(
        self,
        permissions_mask: int = 0o700,
        parents: bool = False,
        exist_ok: bool = False,
    ) -> None:
        """
        Create a directory at this given path. For details, check pathlib.Path.mkdir()
        """
        if parents and not self.parent.exists():
            self.parent.mkdir(
                permissions_mask=permissions_mask, exist_ok=exist_ok, parents=True
            )
        if not self.exists():
            log.info("Creating directory %s", str(self._path))
        self._path.mkdir(mode=permissions_mask, exist_ok=exist_ok)

    def read_text(self, file_size_limit_mb: int, *args, **kwargs) -> str:
        """
        Return the decoded contents of the file as a string.
        Raises an error of the file exceeds the specified size limit.
        For details, check pathlib.Path.read_text()
        """
        self._assert_exists_and_is_file()
        self._assert_file_size_limit(file_size_limit_mb)
        log.info("Reading file %s", self._path)
        return self._path.read_text(*args, **kwargs)

    def write_text(self, *args, **kwargs):
        """
        Open the file pointed to in text mode, write data to it, and close the file.
        """
        if not self.exists():
            self.touch()
        log.info("Writing to file %s", self._path)
        self.path.write_text(*args, **kwargs)

    @contextmanager
    def open(  # noqa: A003
        self,
        mode="r",
        read_file_limit_mb: Optional[int] = None,
        **open_kwargs,
    ):
        """
        Open the file pointed by this path and return a file object, as
        the built-in open() function does.
        If the file is opened for reading, [read_file_limit_kb] parameter must be provided.
        Raises error if the read file exceeds the specified size limit.
        """
        opened_for_reading = "r" in mode
        if opened_for_reading:
            assert (
                read_file_limit_mb is not None
            ), "For reading mode ('r') read_file_limit_mb must be provided"
            self._assert_exists_and_is_file()
            self._assert_file_size_limit(read_file_limit_mb)

        if self.exists():
            self.assert_is_file()
        else:
            self.touch()  # makes sure permissions of freshly-created file are strict

        log.info("Opening file %s in mode '%s'", self._path, mode)
        with self._path.open(mode=mode, **open_kwargs) as fd:
            yield fd
        log.info("Closing file %s", self._path)

    def move(self, destination: Union[Path, str]) -> "SecurePath":
        """Recursively move a file or directory (src) to another location and return the destination.

        If dst is an existing directory or a symlink to a directory, then src is moved inside that directory.
        The destination path in that directory must not already exist.
        """
        destination = Path(destination)
        if destination.is_dir():
            destination = destination / self._path.name
        if destination.exists():
            _raise_file_exists_error(destination)
        log.info("Moving %s to %s", str(self._path), destination.resolve())
        return SecurePath(shutil.move(str(self._path), destination))

    def copy(
        self, destination: Union[Path, str], dirs_exist_ok: bool = False
    ) -> "SecurePath":
        """
        Copy the file/directory into the destination.
        If source is a directory, its whole content is copied recursively.
        Permissions of the copy are limited only to the owner.

        If destination is an existing directory, the copy will be created inside it,
        unless dirs_exist_ok is true and the destination has the same name as this path.

        Otherwise, the copied file/base directory will be renamed to match destination.
        If dirs_exist_ok is false (the default) and dst already exists,
        a FileExistsError is raised. If dirs_exist_ok is true,
        the copying operation will continue if it encounters existing directories,
        and files within the destination tree will be overwritten by corresponding
        files from the src tree.
        """
        self.assert_exists()

        destination = Path(destination)
        if destination.exists():
            if destination.is_dir() and (
                destination.name != self._path.name or self.path.is_file()
            ):
                destination = destination / self._path.name

            if destination.exists():
                if not all([destination.is_dir(), self._path.is_dir(), dirs_exist_ok]):
                    raise FileExistsError(
                        errno.EEXIST, os.strerror(errno.EEXIST), self._path.resolve()
                    )

        def _recursive_check_for_conflicts(src: Path, dst: Path):
            if dst.exists() and not dirs_exist_ok:
                _raise_file_exists_error(dst)
            if dst.is_file() and not src.is_file():
                _raise_not_a_directory_error(dst)
            if dst.is_dir() and not src.is_dir():
                _raise_is_a_directory_error(dst)
            if src.is_dir():
                for child in src.iterdir():
                    _recursive_check_for_conflicts(child, dst / child.name)

        def _recursive_copy(src: SecurePath, dst: SecurePath):
            if src.is_file():
                log.info("Copying file %s into %s", src.path, dst.path)
                if dst.exists():
                    dst.unlink()
                shutil.copyfile(src.path, dst.path)
                dst.restrict_permissions()
            if src.is_dir():
                dst.mkdir(exist_ok=True)
                for child in src.iterdir():
                    _recursive_copy(child, dst / child.name)

        _recursive_check_for_conflicts(self._path, destination)
        _recursive_copy(self, self.__class__(destination))

        return SecurePath(destination)

    def unlink(self, missing_ok=False):
        """
        Remove this file or symbolic link.
        If the path points to a directory, use SecurePath.rmdir() instead.

        Check pathlib.Path.unlink() for details.
        """
        if not self.exists():
            if not missing_ok:
                self.assert_exists()
            return

        self.assert_is_file()
        log.info("Removing file %s", self._path)
        self._path.unlink()

    def rmdir(self, recursive=False, missing_ok=False):
        """
        Remove this directory.
        If the path points to a file, use SecurePath.unlink() instead.

        If path points to a file, NotADirectoryError will be raised.
        If directory does not exist, FileNotFoundError will be raised unless [missing_ok] is True.
        If the directory is not empty, DirectoryNotEmpty will be raised unless [recursive] is True.
        """
        if not self.exists():
            if not missing_ok:
                self.assert_exists()
            return

        self.assert_is_directory()

        if not recursive and any(self._path.iterdir()):
            raise DirectoryIsNotEmptyError(self._path.resolve())

        log.info("Removing directory %s", self._path)
        shutil.rmtree(str(self._path))

    @classmethod
    @contextmanager
    def temporary_directory(cls):
        """
        Creates a temporary directory in the most secure manner possible.
        The directory is readable, writable, and searchable only by the creating user ID.
        Yields SecurePath pointing to the absolute location of created directory.

        Works similarly to tempfile.TemporaryDirectory
        """
        with tempfile.TemporaryDirectory(prefix="snowflake-cli") as tmpdir:
            log.info("Created temporary directory %s", tmpdir)
            yield SecurePath(tmpdir)
            log.info("Removing temporary directory %s", tmpdir)

    def _assert_exists_and_is_file(self) -> None:
        self.assert_exists()
        self.assert_is_file()

    def assert_exists(self) -> None:
        if not self.exists():
            raise FileNotFoundError(
                errno.ENOENT, os.strerror(errno.ENOENT), self._path.resolve()
            )

    def assert_is_file(self) -> None:
        if not self._path.is_file():
            _raise_is_a_directory_error(self._path.resolve())

    def assert_is_directory(self) -> None:
        if not self._path.is_dir():
            _raise_not_a_directory_error(self._path.resolve())

    def _assert_file_size_limit(self, size_limit_in_mb):
        if (
            size_limit_in_mb != UNLIMITED
            and self._path.stat().st_size > size_limit_in_mb * 1024 * 1024
        ):
            raise FileTooLargeError(self._path.resolve(), size_limit_in_mb)

    def rename(self, new_name: Union[str | Path]):
        self._path.rename(new_name)


def _raise_file_exists_error(path: Path):
    raise FileExistsError(errno.EEXIST, os.strerror(errno.EEXIST), path)


def _raise_is_a_directory_error(path: Path):
    raise IsADirectoryError(errno.EISDIR, os.strerror(errno.EISDIR), path)


def _raise_not_a_directory_error(path: Path):
    raise NotADirectoryError(errno.ENOTDIR, os.strerror(errno.ENOTDIR), path)
