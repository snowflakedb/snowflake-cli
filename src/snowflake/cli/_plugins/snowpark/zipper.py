# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import fnmatch
import logging
from pathlib import Path
from typing import Dict, List, Literal
from zipfile import ZIP_DEFLATED, ZipFile

from snowflake.cli.api.artifacts.bundle_map import BundleMap
from snowflake.cli.api.console import cli_console
from snowflake.cli.api.secure_path import SecurePath

log = logging.getLogger(__name__)

IGNORED_FILES = [
    "**/.DS_Store",
    "**/.git",
    "**/.git/*",
    "**/.gitignore",
    "**/.env",
    "**/.env/*",
    "**/.venv",
    "**/.venv/*",
    "**/__pycache__",
    "**/*.zip",
    "**/*.pyc",
    "**/env",
    "**/env/*",
    "**/ENV",
    "**/ENV/*",
    "**/venv",
    "**/venv/*",
    "**/requirements.txt",
    "**/requirements.snowflake.txt",
    "**/requirements.other.txt",
    "**/snowflake.yml",
]


def add_file_to_existing_zip(zip_file: str, file: str):
    """Adds another file to an existing zip file

    Args:
        zip_file (str): The existing zip file
        file (str): The new file to add
    """
    with ZipFile(zip_file, mode="a") as myzip:
        myzip.write(file, Path(file).name)


def zip_dir(
    source: Path | List[Path],
    dest_zip: Path,
    mode: Literal["r", "w", "x", "a"] = "w",
) -> None:

    if not dest_zip.parent.exists():
        SecurePath(dest_zip).parent.mkdir(parents=True)

    if isinstance(source, Path) or isinstance(source, SecurePath):
        source = [source]

    files_to_pack: Dict[Path, List[Path]] = {
        src: list(filter(_to_be_zipped, (f.absolute() for f in src.glob("**/*"))))
        for src in source
    }

    with ZipFile(dest_zip, mode, ZIP_DEFLATED, allowZip64=True) as package_zip:
        for src, files in files_to_pack.items():
            if isinstance(src, SecurePath):
                src = src.path
            for file in files:
                log.debug("Adding %s to %s", file, dest_zip)
                package_zip.write(file, arcname=file.relative_to(src))


def zip_dir_using_bundle_map(
    bundle_map: BundleMap,
    dest_zip: Path,
    mode: Literal["r", "w", "x", "a"] = "w",
) -> None:
    if not dest_zip.parent.exists():
        SecurePath(dest_zip).parent.mkdir(parents=True)

    with ZipFile(dest_zip, mode, ZIP_DEFLATED, allowZip64=True) as package_zip:
        cli_console.step(f"Creating: {dest_zip}")
        for src, _ in bundle_map.all_mappings(expand_directories=True):
            if src.is_file():
                log.debug("Adding %s to %s", src, dest_zip)
                package_zip.write(src, arcname=_path_without_top_level_directory(src))


def _path_without_top_level_directory(path: Path) -> str:
    path_parts = path.parts
    if len(path_parts) > 1:
        return str(Path(*path_parts[1:]))
    return str(path)


def _to_be_zipped(file: Path) -> bool:
    for pattern in IGNORED_FILES:
        # This has to be a string because of fnmatch
        file_as_str = str(file)
        if file_as_str == pattern or fnmatch.fnmatch(file_as_str, pattern):
            return False

    return True
