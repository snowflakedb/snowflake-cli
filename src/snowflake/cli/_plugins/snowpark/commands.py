# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import logging
from collections import defaultdict
from typing import Dict, Optional, Set, Tuple

import typer
from click import ClickException, UsageError
from snowflake.cli._plugins.object.commands import (
    describe as object_describe,
)
from snowflake.cli._plugins.object.commands import (
    drop as object_drop,
)
from snowflake.cli._plugins.object.commands import (
    list_ as object_list,
)
from snowflake.cli._plugins.object.commands import (
    scope_option,
)
from snowflake.cli._plugins.object.manager import ObjectManager
from snowflake.cli._plugins.snowpark import package_utils
from snowflake.cli._plugins.snowpark.common import (
    EntityToImportPathsMapping,
    SnowparkEntities,
    SnowparkObject,
    SnowparkObjectManager,
    StageToArtifactMapping,
    map_path_mapping_to_artifact,
    zip_and_copy_artifacts_to_deploy,
)
from snowflake.cli._plugins.snowpark.package.anaconda_packages import (
    AnacondaPackages,
    AnacondaPackagesManager,
)
from snowflake.cli._plugins.snowpark.package.commands import app as package_app
from snowflake.cli._plugins.snowpark.snowpark_entity_model import (
    FunctionEntityModel,
    ProcedureEntityModel,
)
from snowflake.cli._plugins.snowpark.snowpark_project_paths import (
    SnowparkProjectPaths,
)
from snowflake.cli._plugins.snowpark.snowpark_shared import (
    AllowSharedLibrariesOption,
    IgnoreAnacondaOption,
    IndexUrlOption,
    SkipVersionCheckOption,
)
from snowflake.cli._plugins.snowpark.zipper import zip_dir
from snowflake.cli._plugins.stage.manager import StageManager
from snowflake.cli.api.cli_global_context import (
    get_cli_context,
)
from snowflake.cli.api.commands.decorators import (
    with_project_definition,
)
from snowflake.cli.api.commands.flags import (
    ForceReplaceOption,
    ReplaceOption,
    execution_identifier_argument,
    identifier_argument,
    like_option,
)
from snowflake.cli.api.commands.snow_typer import SnowTyperFactory
from snowflake.cli.api.console import cli_console
from snowflake.cli.api.constants import (
    DEFAULT_SIZE_LIMIT_MB,
)
from snowflake.cli.api.exceptions import (
    SecretsWithoutExternalAccessIntegrationError,
)
from snowflake.cli.api.feature_flags import FeatureFlag
from snowflake.cli.api.identifiers import FQN
from snowflake.cli.api.output.types import (
    CollectionResult,
    CommandResult,
    MessageResult,
    SingleQueryResult,
)
from snowflake.cli.api.project.definition_conversion import (
    convert_project_definition_to_v2,
)
from snowflake.cli.api.project.schemas.project_definition import (
    ProjectDefinition,
    ProjectDefinitionV2,
)
from snowflake.cli.api.secure_path import SecurePath
from snowflake.connector import DictCursor, ProgrammingError
from snowflake.connector.cursor import SnowflakeCursor

log = logging.getLogger(__name__)

app = SnowTyperFactory(
    name="snowpark",
    help="Manages procedures and functions.",
)
app.add_typer(package_app)

ObjectTypeArgument = typer.Argument(
    help="Type of Snowpark object",
    case_sensitive=False,
    show_default=False,
)
IdentifierArgument = identifier_argument(
    "function/procedure",
    example="hello(int, string)",
)
LikeOption = like_option(
    help_example='`list function --like "my%"` lists all functions that begin with “my”',
)


@app.command("deploy", requires_connection=True, require_warehouse=True)
@with_project_definition()
def deploy(
    replace: bool = ReplaceOption(
        help="Replaces procedure or function if there were changes in the definition. It only uploads new and "
        "overwrites existing files, but does not remove any files already on the stage."
    ),
    force_replace: bool = ForceReplaceOption(),
    **options,
) -> CommandResult:
    """
    Deploys procedures and functions defined in project. Deploying the project alters all objects defined in it.
    By default, if any of the objects exist already the commands will fail unless `--replace` flag is provided.
    Required artifacts are deployed before creating functions or procedures. Dependencies are deployed once to
    every stage specified in definitions.
    """
    cli_context = get_cli_context()
    pd = _get_v2_project_definition(cli_context)

    snowpark_entities = get_snowpark_entities(pd)
    project_paths = SnowparkProjectPaths(
        project_root=cli_context.project_root,
    )

    with cli_console.phase("Performing initial validation"):
        if not snowpark_entities:
            raise ClickException(
                "No procedures or functions were specified in the project definition."
            )
        validate_all_artifacts_exists(
            project_paths=project_paths, snowpark_entities=snowpark_entities
        )

    # Validate current state
    with cli_console.phase("Checking remote state"):
        om = ObjectManager()
        _check_if_all_defined_integrations_exists(om, snowpark_entities)
        existing_objects = (
            {}
            if force_replace
            else check_for_existing_objects(om, replace, snowpark_entities)
        )

    with cli_console.phase("Preparing required stages and artifacts"):
        entities_to_imports_map, stages_to_artifact_map = build_artifacts_mappings(
            project_paths=project_paths,
            snowpark_entities=snowpark_entities,
        )

        create_stages_and_upload_artifacts(stages_to_artifact_map)

    # Create snowpark entities
    with cli_console.phase("Creating Snowpark entities"):
        snowpark_manager = SnowparkObjectManager()
        snowflake_dependencies = _read_snowflake_requirements_file(
            project_paths.snowflake_requirements
        )
        deploy_status = []
        for entity in snowpark_entities.values():
            operation_result = snowpark_manager.deploy_entity(
                entity=entity,
                existing_objects=existing_objects,
                snowflake_dependencies=snowflake_dependencies,
                entities_to_artifact_map=entities_to_imports_map,
            )
            deploy_status.append(operation_result)

    return CollectionResult(deploy_status)


def validate_all_artifacts_exists(
    project_paths: SnowparkProjectPaths, snowpark_entities: SnowparkEntities
):
    for key, entity in snowpark_entities.items():
        for artifact in entity.artifacts:
            path = project_paths.get_artifact_dto(artifact).post_build_path
            if not path.exists():
                raise UsageError(
                    f"Artifact {path} required for {entity.type} {key} does not exist."
                )


def check_for_existing_objects(
    om: ObjectManager, replace: bool, snowpark_entities: SnowparkEntities
) -> Dict[str, SnowflakeCursor]:
    existing_objects: Dict[str, SnowflakeCursor] = _find_existing_objects(
        snowpark_entities, om
    )
    if existing_objects and not replace:
        existing_entities = [snowpark_entities[e] for e in existing_objects]
        msg = "Following objects already exists. Consider using --replace.\n"
        msg += "\n".join(f"{e.type}: {e.entity_id}" for e in existing_entities)
        raise ClickException(msg)
    return existing_objects


def build_artifacts_mappings(
    project_paths: SnowparkProjectPaths, snowpark_entities: SnowparkEntities
) -> Tuple[EntityToImportPathsMapping, StageToArtifactMapping]:
    stages_to_artifact_map: StageToArtifactMapping = defaultdict(set)
    entities_to_imports_map: EntityToImportPathsMapping = defaultdict(set)
    for name, entity in snowpark_entities.items():
        stage = entity.stage
        required_artifacts = set()
        for artifact in entity.artifacts:
            artifact_dto = project_paths.get_artifact_dto(artifact)
            required_artifacts.add(artifact_dto)
            entities_to_imports_map[name].add(artifact_dto.import_path(stage))
        stages_to_artifact_map[stage].update(required_artifacts)

        deps_artifact = project_paths.get_dependencies_artifact()
        if deps_artifact.post_build_path.exists():
            stages_to_artifact_map[stage].add(deps_artifact)
            entities_to_imports_map[name].add(deps_artifact.import_path(stage))
    return entities_to_imports_map, stages_to_artifact_map


def create_stages_and_upload_artifacts(stages_to_artifact_map: StageToArtifactMapping):
    stage_manager = StageManager()
    for stage, artifacts in stages_to_artifact_map.items():
        cli_console.step(f"Creating (if not exists) stage: {stage}")
        stage = FQN.from_stage(stage).using_context()
        stage_manager.create(fqn=stage, comment="deployments managed by Snowflake CLI")
        for artifact in artifacts:
            post_build_path = artifact.post_build_path
            cli_console.step(
                f"Uploading {post_build_path.name} to {artifact.upload_path(stage)}"
            )
            stage_manager.put(
                local_path=post_build_path,
                stage_path=artifact.upload_path(stage),
                overwrite=True,
            )


def _find_existing_objects(
    objects: SnowparkEntities,
    om: ObjectManager,
) -> Dict[str, SnowflakeCursor]:
    existing_objects = {}
    for entity_id, entity in objects.items():
        identifier = entity.udf_sproc_identifier.identifier_with_arg_types
        try:
            current_state = om.describe(
                object_type=entity.type,
                fqn=FQN.from_string(identifier),
            )
            existing_objects[entity_id] = current_state
        except ProgrammingError:
            pass
    return existing_objects


def _check_if_all_defined_integrations_exists(
    om: ObjectManager,
    snowpark_entities: SnowparkEntities,
):
    existing_integrations = {
        i["name"].lower()
        for i in om.show(object_type="integration", cursor_class=DictCursor, like=None)
        if i["type"] == "EXTERNAL_ACCESS"
    }
    declared_integration: Set[str] = set()
    for object_definition in snowpark_entities.values():
        external_access_integrations = {
            s.lower() for s in object_definition.external_access_integrations
        }
        secrets = [s.lower() for s in object_definition.secrets]

        if not external_access_integrations and secrets:
            raise SecretsWithoutExternalAccessIntegrationError(object_definition.fqn)

        declared_integration = declared_integration | external_access_integrations

    missing = declared_integration - existing_integrations
    if missing:
        raise ClickException(
            f"Following external access integration does not exists in Snowflake: {', '.join(missing)}"
        )


def _read_snowflake_requirements_file(file_path: SecurePath):
    if not file_path.exists():
        return []
    return file_path.read_text(file_size_limit_mb=DEFAULT_SIZE_LIMIT_MB).splitlines()


@app.command("build", requires_connection=True)
@with_project_definition()
def build(
    ignore_anaconda: bool = IgnoreAnacondaOption,
    allow_shared_libraries: bool = AllowSharedLibrariesOption,
    index_url: Optional[str] = IndexUrlOption,
    skip_version_check: bool = SkipVersionCheckOption,
    **options,
) -> CommandResult:
    """
    Builds artifacts required for the Snowpark project. The artifacts can be used by `deploy` command.
    For each directory in artifacts a .zip file is created. All non-anaconda dependencies are packaged in
    dependencies.zip file.
    """
    cli_context = get_cli_context()
    pd = _get_v2_project_definition(cli_context)

    project_paths = SnowparkProjectPaths(
        project_root=cli_context.project_root,
    )

    anaconda_packages_manager = AnacondaPackagesManager()

    # Clean up bundle root
    project_paths.remove_up_bundle_root()

    # Resolve dependencies
    if project_paths.requirements.exists():
        with (
            cli_console.phase("Resolving dependencies from requirements.txt"),
            SecurePath.temporary_directory() as temp_deps_dir,
        ):
            requirements = package_utils.parse_requirements(
                requirements_file=project_paths.requirements,
            )
            anaconda_packages = (
                AnacondaPackages.empty()
                if ignore_anaconda
                else anaconda_packages_manager.find_packages_available_in_snowflake_anaconda()
            )
            download_result = package_utils.download_unavailable_packages(
                requirements=requirements,
                target_dir=temp_deps_dir,
                anaconda_packages=anaconda_packages,
                skip_version_check=skip_version_check,
                pip_index_url=index_url,
            )

            log.info("Checking to see if packages have shared (.so/.dll) libraries...")
            if package_utils.detect_and_log_shared_libraries(
                download_result.downloaded_packages_details
            ):
                if not allow_shared_libraries:
                    raise ClickException(
                        "Some packages contain shared (.so/.dll) libraries. "
                        "Try again with --allow-shared-libraries."
                    )
            if download_result.anaconda_packages:
                anaconda_packages.write_requirements_file_in_snowflake_format(  # type: ignore
                    file_path=project_paths.snowflake_requirements,
                    requirements=download_result.anaconda_packages,
                )

            if any(temp_deps_dir.path.iterdir()):
                dep_artifact = project_paths.get_dependencies_artifact()
                cli_console.step(f"Creating {dep_artifact.path.name}")
                zip_dir(
                    source=temp_deps_dir.path,
                    dest_zip=dep_artifact.post_build_path,
                )
            else:
                cli_console.step(f"No external dependencies.")

    artifacts = set()
    with cli_console.phase("Preparing artifacts for source code"):
        for entity in get_snowpark_entities(pd).values():
            artifacts.update(
                map_path_mapping_to_artifact(project_paths, entity.artifacts)
            )

        if FeatureFlag.ENABLE_SNOWPARK_GLOB_SUPPORT.is_enabled():
            zip_and_copy_artifacts_to_deploy(artifacts, project_paths.bundle_root)
        else:
            for artifact in artifacts:
                artifact.build()

    return MessageResult(f"Build done.")


def get_snowpark_entities(
    pd: ProjectDefinition,
) -> Dict[str, ProcedureEntityModel | FunctionEntityModel]:
    procedures: Dict[str, ProcedureEntityModel] = pd.get_entities_by_type("procedure")
    functions: Dict[str, FunctionEntityModel] = pd.get_entities_by_type("function")
    snowpark_entities = {**procedures, **functions}
    return snowpark_entities


@app.command("execute", requires_connection=True)
def execute(
    object_type: SnowparkObject = ObjectTypeArgument,
    execution_identifier: str = execution_identifier_argument(
        "procedure/function", "hello(1, 'world')"
    ),
    **options,
) -> CommandResult:
    """Executes a procedure or function in a specified environment."""
    cursor = SnowparkObjectManager().execute(
        execution_identifier=execution_identifier, object_type=object_type
    )
    return SingleQueryResult(cursor)


@app.command("list", requires_connection=True)
def list_(
    object_type: SnowparkObject = ObjectTypeArgument,
    like: str = LikeOption,
    scope: Tuple[str, str] = scope_option(
        help_example="`list function --in database my_db`"
    ),
    **options,
):
    """Lists all available procedures or functions."""
    return object_list(object_type=object_type.value, like=like, scope=scope, **options)


@app.command("drop", requires_connection=True)
def drop(
    object_type: SnowparkObject = ObjectTypeArgument,
    identifier: FQN = IdentifierArgument,
    **options,
):
    """Drop procedure or function."""
    return object_drop(object_type=object_type.value, object_name=identifier, **options)


@app.command("describe", requires_connection=True)
def describe(
    object_type: SnowparkObject = ObjectTypeArgument,
    identifier: FQN = IdentifierArgument,
    **options,
):
    """Provides description of a procedure or function."""
    return object_describe(
        object_type=object_type.value, object_name=identifier, **options
    )


def _get_v2_project_definition(cli_context) -> ProjectDefinitionV2:
    pd = cli_context.project_definition
    if not pd.meets_version_requirement("2"):
        pd = convert_project_definition_to_v2(cli_context.project_root, pd)
    return pd
