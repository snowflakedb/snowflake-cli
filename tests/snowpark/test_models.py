# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import pytest
from snowflake.cli._plugins.snowpark.models import (
    Requirement,
    WheelMetadata,
    get_package_name,
)


@pytest.mark.parametrize(
    "line,name,extras",
    [
        ("ipython ; extra == 'docs'", "ipython", ["docs"]),
        ("foo", "foo", []),
        ("pytest ; extra == 'tests'", "pytest", ["tests"]),
    ],
)
def test_requirement_is_parsed_correctly(line, name, extras):
    result = Requirement.parse_line(line)

    assert result.name == name
    assert result.extras == extras


@pytest.mark.parametrize(
    "line,name",
    [
        (
            "git+https://github.com/sfc-gh-turbaszek/dummy-pkg-for-tests",
            "dummy-pkg-for-tests",
        ),
        (
            "git+https://github.com/sfc-gh-turbaszek/dummy-pkg-for-tests@foo",
            "dummy-pkg-for-tests",
        ),
        (
            "git+https://github.com/sfc-gh-turbaszek/dummy-pkg-for-tests@0123456789abcdef0123456789abcdef01234567",
            "dummy-pkg-for-tests",
        ),
        ("foo.zip", "foo"),
        ("package", "package"),
        ("package.zip", "package"),
        ("git+https://github.com/snowflakedb/snowflake-cli/", "snowflake-cli"),
        (
            "git+https://github.com/snowflakedb/snowflake-cli.git/@snow-123456-fix",
            "snowflake-cli",
        ),
    ],
)
def test_get_package_name(line, name):
    assert get_package_name(line) == name


def test_wheel_metadata_parsing(test_root_path):
    from snowflake.cli._plugins.snowpark.zipper import zip_dir
    from snowflake.cli.api.secure_path import SecurePath

    with SecurePath.temporary_directory() as tmpdir:
        wheel_path = tmpdir / "Zendesk-1.1.1-py3-none-any.whl"

        # prepare .whl package
        package_dir = tmpdir / "ZendeskWhl"
        package_dir.mkdir()
        package_src = (
            SecurePath(test_root_path) / "test_data" / "local_packages" / ".packages"
        )
        for srcdir in ["zendesk", "Zendesk-1.1.1.dist-info"]:
            (package_src / srcdir).copy(package_dir.path)
        zip_dir(source=package_dir.path, dest_zip=wheel_path.path)

        # check metadata
        meta = WheelMetadata.from_wheel(wheel_path.path)
        assert meta.name == "zendesk"
        assert meta.wheel_path == wheel_path.path
        assert meta.dependencies == ["httplib2", "simplejson"]


def test_raise_error_when_artifact_contains_asterix(
    runner, project_directory, alter_snowflake_yml, os_agnostic_snapshot
):
    with project_directory("glob_patterns") as tmp_dir:
        alter_snowflake_yml(
            tmp_dir / "snowflake.yml", "entities.hello_procedure.artifacts", ["src/*"]
        )

        result = runner.invoke(["snowpark", "build"])

        assert result.exit_code == 1
        assert result.output == os_agnostic_snapshot
