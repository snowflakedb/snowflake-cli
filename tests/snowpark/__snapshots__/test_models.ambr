# serializer version: 1
# name: test_raise_error_when_artifact_contains_asterix
  '''
  +- Error ----------------------------------------------------------------------+
  | During evaluation of DefinitionV20 in project definition following errors    |
  | were encountered:                                                            |
  | For field entities.hello_procedure.procedure.artifacts you provided          |
  | '['src/*']'. This caused: Value error, If you want to use glob patterns in   |
  | artifacts, you need to enable the Snowpark new build feature flag            |
  | (enable_snowpark_glob_support=true)                                          |
  +------------------------------------------------------------------------------+
  
  '''
# ---
